src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs

```diff
@@ -131,13 +201,128 @@ private static bool ApplyChangesFromTC25ThroughTC28(JObject sarifLog)
         }
 
+        private static JArray ConstructLogicalLocationsArray(
+            JObject logicalLocations,
+            Dictionary<JObject, int> jObjectToIndexMap,
+            Dictionary<LogicalLocation, int> logicalLocationToIndexMap,
+            out Dictionary<string, int> fullyQualifiedLogicalNameToIndexMap)
+        {
+            fullyQualifiedLogicalNameToIndexMap = null;
+
+            if (logicalLocations == null) { return null; }
+
+            logicalLocationToIndexMap = new Dictionary<LogicalLocation, int>(LogicalLocation.ValueComparer);
+
+            fullyQualifiedLogicalNameToIndexMap = new Dictionary<string, int>();
+
+            foreach (JProperty logicalLocationEntry in logicalLocations.Properties())
+            {
+                // This condition may occur if we've already procssed a parent logical
+                // location that happened to be enumerated after one of its children
+                if (fullyQualifiedLogicalNameToIndexMap.ContainsKey(logicalLocationEntry.Name)) { continue; }
+
+                fullyQualifiedLogicalNameToIndexMap = CreateLogicalLocationArrayEntry(
+                    logicalLocations,
+                    logicalLocationEntry.Name, 
+                    (JObject)logicalLocationEntry.Value,
+                    logicalLocationToIndexMap, 
+                    jObjectToIndexMap,
+                    fullyQualifiedLogicalNameToIndexMap);
+            }
+
+            var updatedArrayElements = new JObject[jObjectToIndexMap.Count];
+
+            foreach (KeyValuePair<JObject, int> keyValuePair in jObjectToIndexMap)
+            {
+                int index = keyValuePair.Value;
+                JObject updatedLogicalLocation = keyValuePair.Key;
+                updatedArrayElements[index] = updatedLogicalLocation;
+            }
+
+            return new JArray(updatedArrayElements);
+        }
+
+        private static Dictionary<string, int> CreateLogicalLocationArrayEntry(
+            JObject logicalLocationsDictionary,
+            string keyName,
+            JObject logicalLocation,
+            Dictionary<LogicalLocation, int> logicalLocationToIndexMap,
+            Dictionary<JObject, int> jObjectToIndexMap,
+            Dictionary<string, int> keyToIndexMap)
+        {
+            int index;
+            keyToIndexMap = keyToIndexMap ?? new Dictionary<string, int>();
+
+            string fullyQualifiedName = keyName;
+
+            // Parent key is no longer relevant and will be removed.
+            // We require 
+            string parentKey = (string)logicalLocation["parentKey"];
+
+            if (parentKey == null)
+            {
+                // No parent key? We are a root location. 
+                logicalLocation["parentIndex"] = -1;
+            }
+            else
+            {
+                // Have we processed our parent yet? If so, retrieve the parentIndex and we're done
+                if (!keyToIndexMap.TryGetValue(parentKey, out index))
```

Comment:
>index [](start = 62, length = 5)

Rather than reusing `index` (which forces you to declare `index` at the top of the method, increasing the reader's cognitive load), consider using a local with a more specific name:
```
if (!keyToIndexMap.TryGetValue(parentKey, out int parentIndex))
```
... using the C# 7 out parameter declaration feature.